numpy/lib/function_base.py

```diff
@@ -1904,10 +1927,89 @@ def diff(a, n=1, axis=-1):
     slice1 = tuple(slice1)
     slice2 = tuple(slice2)
     if n > 1:
-        return diff(a[slice1]-a[slice2], n-1, axis=axis)
-    else:
+        return diff(a[slice1]-a[slice2], n-1, axis=axis, 
+                    prepend=prepend, append=append)
```

Comment:
There was a remark earlier about what to do with higher-order diffs and `"first"`, which I think was lost. The easy fix is just to disallow these shorthands for `n>1`, and force people to be explicit about `diff(diff(x), prepend='first')` vs ``diff(diff(x, prepend='first'))`.

Otherwise, I think you need to make `np.cumsum(np.cumsum(...))` do the right thing for `n=2`, etc.